File: prime.py
def main():
    import sys
    sys.setrecursionlimit(1 << 25)
    N = 10**6 + 1
    min_prime = [0] * N
    primes = []
    for i in range(2, N):
        if min_prime[i] == 0:
            min_prime[i] = i
            primes.append(i)
        for p in primes:
            if p * i >= N:
                break
            if min_prime[i] < p:
                continue
            min_prime[p * i] = p
    prime_index = {p: idx for idx, p in enumerate(primes)}
    
    T = int(sys.stdin.readline())
    for _ in range(T):
        a, b = map(int, sys.stdin.readline().split())
        count_map = {}
        for num in range(a, b + 1):
            temp = num
            factors = set()
            while temp > 1:
                factor = min_prime[temp]
                temp //= min_prime[temp]
                factors.add(factor)
            if factors:
                indexed_factors = tuple(sorted(prime_index[f] for f in factors))
                count_map[indexed_factors] = count_map.get(indexed_factors, 0) + 1
        total = 0
        for count in count_map.values():
            if count >= 2:
                total += count * (count - 1) // 2
        print(total)

if __name__ == "__main__":
    main()

File: prime.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX_N = 1e6 + 1;

int prime_index[MAX_N];
int min_prime[MAX_N];
struct VectorHash {
    size_t operator()(const vector<int>& v) const {
        size_t hash = 0;
        for (int num : v) {
            hash ^= hash * 31 + num;
        }
        return hash;
    }
};
void preprocess_sieve() {
    vector<int> primes;

    for (int i = 2; i < MAX_N; ++i) {
        if (min_prime[i] == 0) { 
            min_prime[i] = i;
            primes.push_back(i);
        }
        for (int p : primes) {
            if (p * i >= MAX_N) break;
            if (min_prime[i] < p) break;
            min_prime[p * i] = p;
        }
    }

    for (size_t idx = 0; idx < primes.size(); ++idx) {
        prime_index[primes[idx]] = idx;
    }
}

int factorize(int num) {
    int ans = 1;
    while (num > 1) {
        int prime = min_prime[num];
        ans *= prime;
        while (num % prime == 0) { 
            num /= prime;
        }
    }

    return ans;
}
int p[(int) 1e6 + 5];
#include<ext/pb_ds/assoc_container.hpp>
using namespace __gnu_pbds;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    preprocess_sieve();
    for(int i = 1; i <= 1e6; i++) {
        p[i] = factorize(i);
    }
    int T;
    cin >> T;
    while (T--) {
        int a, b;
        cin >> a >> b;

        gp_hash_table<int, int> count_map;

        for (int num = a; num <= b; ++num) {
            count_map[p[num]]++;
        }

        long long total = 0;
        for  (auto [key, count] : count_map) {
            if (count >= 2) {
                total += static_cast<long long>(count) * (count - 1) / 2;
            }
        }

        cout << total << '\n';
    }

    return 0;
}